txpipe/utils/theory.py

```diff
@@ -51,111 +51,206 @@ def ccl_read_yaml(filename, **kwargs):
-    theory_cl: dict{str:array}
-        theory c_ell for each pair (i,j,k) where k is one of
-        SHEAR_SHEAR = 0, SHEAR_POS = 1, POS_POS = 2
+    sacc_theory: sacc.Sacc
+        A copy of the input sacc_data but with values
+        replaced with theory predictions.
+    """
+    import firecrown.likelihood.gauss_family.statistic.source.weak_lensing as wl
+    import firecrown.likelihood.gauss_family.statistic.source.number_counts as nc
+    from firecrown.likelihood.gauss_family.statistic.two_point import TwoPoint
+    import sacc
+    import pathlib
+
+
+    if isinstance(sacc_data, (str, pathlib.Path)):
+        sacc_data = sacc.Sacc.load_fits(sacc_data)
+    else:
+        sacc_data = sacc_data.copy()
+
+    if smooth:
+        smooth_sacc_nz(sacc_data)
+
+    if len(sacc_data.data) == 0:
+        if ell_values is None:
+            raise ValueError("Supplied an empty sacc file but no ell_values to fill it")
+        fill_empty_sacc(sacc_data, ell_values)
+
+
+
+    if isinstance(bias, float):
+        bias = compute_fiducial_bias(bias, sacc_data, cosmo)
+    elif bias is None:
+        bias = [1 for _ in sacc_data.tracers]
+        print("Using bias = 1")
+    else:
+        print("Bias = ", bias)
+
+    # Creating sources. This will presumably be done inside
+    # firecrown at some point
+    sources = {}
+    for tracer_name in sacc_data.tracers.keys():
+        if tracer_name.startswith("source"):
+            tr = wl.WeakLensing(sacc_tracer=tracer_name)
+        elif tracer_name.startswith("lens"):
+            tr = nc.NumberCounts(sacc_tracer=tracer_name)
+            i = int(tracer_name.split('_')[1])
+            tr.bias = bias[i]
+            tr.mag_bias = 0.0
+        else:
+            raise ValueError(f"Unknown tracer in sacc file, non-3x2pt: {tracer}")
+        sources[tracer_name] = tr
 
+    # Making predictions using the TwoPoint statistic object
+    # from firecrown
+    theory = {}
+    for dt in sacc_data.get_data_types():
+        for t1, t2 in sacc_data.get_tracer_combinations(dt):
+            try:
+                stat = TwoPoint(
+                        source0=sources[t1],
+                        source1=sources[t2],
+                        sacc_data_type=dt,
+                )
+            except ValueError as err:
+                # B modes and things like that can be in the files
+                # but are not supported
+                if "is not supported" in str(err):
+                    print(f"Setting bin {t1}, {t2} for {dt} to zero")
+                    n = sacc_data.indices(dt, (t1,t2)).size
+                    theory[dt, t1, t2] = np.zeros(n)
+                    continue
+                else:
+                    raise
+
+            # Read statistic information from the sacc file
+            stat.read(sacc_data)
+
+            # The first output is the data vector, and
+            # we keep the second which is the theory
+            _, theory[dt, t1, t2] = stat.compute(cosmo)
```

Comment:
I'd avoid using this low-level API (`stat.compute`) for now.
There's a firecrown PR in the works to add supports non-linear models (PT, halo model, etc). This changes the `cosmo` object here to a firecrown `Cosmology` object that carries around various CCL calculators along the CCL cosmology object. For this to work, the top level of the firecrown pipeline (the likelihood objects) now does some setup stuff too. TL;DR: this API will break soon and also lock you out of more sophisticated models in firecrown.

So I'd set up a dummy likelihood here,  use that to compute the theory predictions, and then extract the predictions.
There's a firecrown issue (https://github.com/LSSTDESC/firecrown/issues/195) to make this less awkward.

Reply:
Updated to use a dummy likelihood.